### PR TITLE
Builder pattern, XPath parsing, deprecate insecure, change initialization

### LINKED ISSUE #5: Parse XPath filters to Path
Currently ignores (unwillingly) XPath filters which could otherwise be expressed in Path and will ultimately break if filters are provided.

### PR DESCRIPTION
**Pull request by remingtonc**
**EDIT**: The scope of this pull request is getting out of hand, but it's all good stuff. Think it's ready to merge. Bumping to `v1.0.0` due to major changes.

Refactors the internal codebase to be more extensible, particularly with respect to potential deviations in how XPaths are handled and with regard to security, and includes XPath parsing which can handle filters.

* Usage is not simply `XRClient(...)`, changed to builder pattern with `ClientBuilder`. I am especially interested in your opinions around how this looks.

Original:
```python
client = XRClient('127.0.0.1:57400', 'demo', 'demo', credentials='ems.pem', tls_server_override='ems.cisco.com', credentials_from_file=True)
capabilities = client.capabilities()
print(capabilities)
```

New:
```python
from cisco_gnmi import ClientBuilder
client = ClientBuilder(
    '127.0.0.1:9339'
).set_os(
    'IOS-XR'
).set_secure_from_file(
    'ems.pem'
).set_ssl_target_override(
    'ems.cisco.com'
).set_call_authentication(
    'admin',
    'its_a_secret'   
).construct()
capabilities = client.capabilities()
print(capabilities)
```

It's very difficult to mask direct interaction with the gRPC libraries without this sort of initialization, or having tons of optional init arguments. It becomes important once security gets involved. The work done here I will likely attempt to percolate into our other libraries too, so it sets a small precedence on how we initialize clients.

* Wrapper classes now override `parse_xpath_to_gnmi_path` in order to appropriately handle origins etc.
* Default gNMI port changed to 9339 per [reserved_port.md](https://github.com/openconfig/gnoi/blob/master/docs/reserved_port.md).

**EDITed past here**

* Almost forgot, also revamps XPath parsing to actually parse filters etc. correctly into gNMI paths. Borrows the xml module's XPath regex and works around its results. This was very critical.
* Deprecates insecure. This is expressly against gNMI spec and is pretty trivial with the ability to get the certificate from the target.
* Uses channel credentials instead of passing metadata with every call.
* `Client` and child classes simply initialize on top of a gRPC channel instead of any connectivity logic in the constructor.

This work was derived from adding an IOS XE wrapper, but having challenges with SSL & friends. This work in the meantime is still worth discussing/merging.

Resolves #5 
Basis for #10 and #9

### CODE REVIEW

**Comment by remingtonc** [comment]
ping

File: src/cisco_gnmi/base.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+        )
+        self.channel = grpc.secure_channel(
+            self.target_netloc.netloc, self.channel_creds, self.channel_options
+        )
+        if self.service_class:
+            self.apply_service(self.service_class)
+        return self
+
+    def as_insecure(self, channel_options=None, compression=None):
+        logging.warning(
```
Comment (skkumaravel) [inline src/cisco_gnmi/base.py]: Needs a function comment

**Review by skkumaravel — CHANGES_REQUESTED** [review]
Looks good, couple comments:

1. Interesting way to get the CA name, I like it! Is that considered secure, or should that be a flag that is allowed to be passed? The reason I am asking is because by passing the CA name and the pem file, you are ensuring that the pem file is the right pem file and didn't get tampered with.

2. You should also add an example of insecure and secure with file name. Little confused on how to use all the different ways to do an init by just reading the code, especially the secure one.

**Comment by remingtonc** [comment]
@skkumaravel Addressed comments, README.md has more detail now. Regarding the CA name - in the case of `ems.pem` we still need to set the override as `ems.cisco.com` regardless of where it is loaded - so yes it is still useful even locally. However I am also not an expert in this situation.

File: README.md [hunk 2]
```diff
@@ -19,6 +19,148 @@ It is *highly* recommended that users of the library learn [Google Protocol Buff
...
+    from_file=True
+).with_authentication(
+    'admin',
+    'its_a_secret'
+)
+```
+
+
 ### XRClient
 `XRClient` inherets from `Client` and provides several wrapper methods which aid with IOS XR-specific behaviors of the gNMI implementation. These are `delete_xpaths`, `get_xpaths`, `set_json`, and `subscribe_xpaths`. These methods make several assumptions about what kind of information will be supplied to them in order to simplify usage of the gNMI RPCs.
```
Comment (ziausyed) [inline README.md]: Do you want to specify what assumptions are made?

**Review by ziausyed — COMMENTED** [review]
very cool changes.

File: README.md (hunk 2 above)
Comment (remingtonc) [reply README.md]: Not necessarily, the method doc and parameters themselves kind of embody what those assumptions are.

File: README.md [hunk 3]
```diff
@@ -19,6 +19,148 @@ It is *highly* recommended that users of the library learn [Google Protocol Buff
...
+
+Using an owned root certificate on the filesystem:
+
+```python
+from cisco_gnmi import Client
+
+client = Client(
+    '127.0.0.1:9339'
+).as_secure(
+    'ems.pem',
```
Comment (skkumaravel) [inline README.md]: In this example, how is the CA determined?

**Review by skkumaravel — APPROVED** [review]
README helps a lot! Looks really good now, just had the one question..

File: README.md (hunk 3 above)
Comment (remingtonc) [reply README.md]: I quite honestly do not know. This is a good exercise in security.

**Review by brh55 — APPROVED** [review]

**Comment by remingtonc** [comment]
@brh55, despite approving, brought up that what was presented wasn't exactly a builder pattern. He was correct, and I almost brushed it off, but rewrote and now is drastically more builder...and pushed in some extra changes while rolling...please re-review.

PS I'm sorry about wild PR scope creep.

File: README.md (hunk 2 above)
Comment (remingtonc) [reply README.md]: Called out the assumptions being in the doc.

**Comment by remingtonc** [comment]
Resolves #21

**Pull request merged**